sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala

```diff
@@ -401,6 +405,29 @@ object ShuffleExchangeExec {
     dependency
   }
 
+  /**
+   * Checks if the shuffle partitioning contains indeterminate expression/reference.
+   */
+  private def isPartitioningIndeterminate(partitioning: Partitioning, plan: SparkPlan): Boolean = {
```

Comment:
I think we need to build a framework to properly propagate the column-level nondeterministic information. This function looks quite fragile

Reply:
For example, `Filter(rand_cond, Project(a, b, c, ...))`. I think all the columns are nondeterministic after `Filter`, even though attributes `a`, `b` and `c` are deterministic.

Reply:
You mean the QueryPlan's `deterministic`?  https://github.com/apache/spark/pull/34470

Reply:
that's plan level, not column level. We need something more fine-grained

Reply:
ok, I'll first try to add a column level nondeterministic information before this pr